File: 1823/solution.ts
function findTheWinner(n: number, k: number): number {
  const circle: number[] = [];
  for (let i = 1; i <= n; i++) {
    circle.push(i);
  }
  while (circle.length > 1) {
    for (let i = 0; i < k - 1; i++) {
      circle.push(circle.shift()!);
    }
    circle.shift();
  }
  return circle.shift()!;
}

File: 1823/solution.java
import java.util.ArrayDeque;

class Solution {
    public int findTheWinner(int n, int k) {
        ArrayDeque<Integer> circle = new ArrayDeque<>();
        for (int i = 1; i <= n; i++) {
            circle.addLast(i);
        }
        while (circle.size() > 1) {
            for (int i = 0; i < k - 1; i++) {
                circle.addLast(circle.pollFirst());
            }
            circle.pollFirst();
        }
        return circle.getFirst();
    }
}

// time O(n*k)
// space O(n)